lib/proxy_api/bmc.rb

```diff
@@ -49,9 +51,11 @@ def power(args)
       case args[:action]
       when "on?", "off?", "status"
         args[:action].chop! if args[:action].include?('?')
-        response = parse(get(bmc_url_for('power', args[:action]), args))
+        provider_query = @provider ? "?bmc_provider=#{@provider}" : ''
+        response = parse(get(bmc_url_for('power', args[:action]) + provider_query, args))
```

Comment:
Should `bmc_url_for` take an optional provider argument instead? That feels like a better implementation. The URL might already include an argument and a simple string concat with `?` will fail in that case.

Reply:
In general I don't know what the structural fix is. There are a lot of places in this part of the code where awkward assumptions are made and I don't feel like I have enough sense of what is required, philosophically, to have wanted to do anything more than a pragmatic fix. My gut feeling was that in this UI code, at this time, we're not ever going to see a ? in the url returned from bmc_url_for().

Reply:
...but I'm open to direction.

Reply:
I did pursue having the provider data sent in the body (as done in the documentation examples), but I couldn't get it to survive through the get() method call. The best I managed was to get it to show up in an awkward way in the HTTP headers, which felt wrong.

Reply:
This isn't really UI code. The BMC API is only called by Foreman to connect to the Foreman Proxy. Clients to Foreman shouldn't ever talk to the Foreman Proxy.

My suggestion would be to change https://github.com/theforeman/foreman/blob/203cb8865eb313a0b0269b0d5d7abe3b617e3bce/lib/proxy_api/bmc.rb#L102-L109 to add a third parameter (`provider = nil`) and handle it there. Now that I see it's actually a private API you can also generate URLs with `?provider=#{@provider}`. That would solve it in one place instead of 3.

Reply:
On further reflection:

I'm not sure changing bmc_url_for() to add a provider argument is a good strategy. I think that would require bmc_url_for() to know that the returned url would be used with the GET method or not, as with POST or PUT, it'd be part of the request body instead.

Reply:
Hmm, fair point. I think the main issue here is that it's a bad API design on the REST API. It should have been part of the path (like `/#{@provider}/#{@target}/chassis/#{controller}/#{action}`) which would have prevented all this ugly workarounds on the consumer side, but that's not going to change soon.

Reply:
Are we good here, @ekohl ? I am fine as long as this works, which I am going to test in a bit.

Reply:
Yes, we're not going to fix the ugly API easily.

Reply:
Can't this be solved like this:
```suggestion
        args[:bmc_provider] = @provider if @provider
        response = parse(get(bmc_url_for('power', args[:action]), args))
```

Reply:
I couldn't get any "normal" args to survive through the call chain to HTTP GET. I got more than neck deep into it and (IIRC) it was the result of implementation details in the third-party HTTP library. Way beyond my scope.

Reply:
That said, I did most of this work in Foreman 1.24, and my virtualized dev environment evolved pretty sketchily from there. It's possible the library's been fixed?